### PR TITLE
Keysend bLIP

### PR DESCRIPTION
**Pull request by valentinewallace**
Currently rebased on #884 

The portions regarding the keysend feature bit and MPP probably need more iteration, but wanted to get a draft out.

### CODE REVIEW

File: blips/blip-0002.mediawiki [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  randomly generated and cryptographically-secure 32-byte value that serves as
+  the HTLC payment preimage
+* MUST NOT set a payment secret
+* SHOULD only send payments to nodes advertising feature bit 55 in the node
+  context starting June 1, 2022
+
+Receiver:
+* if completing the payment, MUST fulfill the HTLC using the TLV-provided
+  payment preimage
+* if failing the payment, SHOULD error with `PERM|15
```
Comment (TheBlueMatt) [inline blips/blip-0002.mediawiki:40]: Should specify reason -> value, not just "if you fail for any reason, use this code".

File: blips/blip-0002.mediawiki [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+Sender:
+* MUST include a TLV record keyed by type `5482373484` with a TLV value of a
+  randomly generated and cryptographically-secure 32-byte value that serves as
+  the HTLC payment preimage
+* MUST NOT set a payment secret
+* SHOULD only send payments to nodes advertising feature bit 55 in the node
+  context starting June 1, 2022
+
+Receiver:
+* if completing the payment, MUST fulfill the HTLC using the TLV-provided
```
Comment (TheBlueMatt) [inline blips/blip-0002.mediawiki:38]: Does this need to be specified?

File: blips/blip-0002.mediawiki [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  payment preimage
+* if failing the payment, SHOULD error with `PERM|15
+  incorrect_or_unknown_payment_details`.
+* MUST advertise feature bit 55 in the node context only starting June 1, 2022
```
Comment (TheBlueMatt) [inline blips/blip-0002.mediawiki:42]: I think we should start advertising now.

File: blips/blip-0002.mediawiki [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+==Rationale==
+
+Design decisions for keysend were largely made by the original lnd keysend
+implementation (e.g. the choice of `5482373484` for the TLV type).
+
+* The TLV type `5482373484` was chosen at random
+* The feature bit is currently not supported by all implementations, but should
+be supported after June 1, 2022. The purpose of this lead time is to give
+implementations the chance to implement the feature without being
+bLIP-incompatible in the meantime. The rationale behind having a feature bit is
```
Comment (TheBlueMatt) [inline blips/blip-0002.mediawiki:76]: I don't think being "bLIP-incompatible" is an issue, so much as refusing to send a keysend payment to a node that would otherwise support it.

File: blips/blip-0002.mediawiki [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+that it's nicer to opt-in to receiving keysend payments: without a feature bit,
+the only way for senders to explicitly know that receivers support keysend is
+by attempting a keysend payment and seeing whether or not it fails.
+* The rationale for not supporting basic MPP is that it doesn't make much sense for
```
Comment (TheBlueMatt) [inline blips/blip-0002.mediawiki:80]: I mean I think the bigger rationale is that implementations currently dont support it, but, yea, this too :)

**Review by TheBlueMatt — COMMENTED** [review]
Thanks for writing this up, good to have a spec for this.

File: blips/blip-0002.mediawiki [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+Sender:
+* MUST include a TLV record keyed by type `5482373484` with a TLV value of a
+  randomly generated and cryptographically-secure 32-byte value that serves as
+  the HTLC payment preimage
+* MUST NOT set a payment secret
+* SHOULD only send payments to nodes advertising feature bit 55 in the node
+  context starting June 1, 2022
+
+Receiver:
```
Comment (ariard) [inline blips/blip-0002.mediawiki]: MUST reject the keysend if the preimage is not matching the payment hash ?
MUST reject the keysend if a `payment data` field is present ?

File: blips/blip-0002.mediawiki [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+* if completing the payment, MUST fulfill the HTLC using the TLV-provided
+  payment preimage
+* if failing the payment, SHOULD error with `PERM|15
+  incorrect_or_unknown_payment_details`.
+* MUST advertise feature bit 55 in the node context only starting June 1, 2022
+
```
Comment (ariard) [inline blips/blip-0002.mediawiki]: I think you should include somewhere that TLV type 8 is from now occupied in the `tlv_payload` by `keysend_preimage` ?

File: blips/blip-0002.mediawiki (hunk 6 above)
Comment (ariard) [reply blips/blip-0002.mediawiki]: I think you can even the deanonymization attack in the lack of preimage verification in the rational section.

File: blips/blip-0002.mediawiki [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+As previously mentioned, keysend also has the advantage of already being
+supported by major implementations. So regardless of whether it is the best
+solution to spontaneous payments, it's a good idea to have some form of
+official documentation for it. *However*, because of the drawbacks of keysend
+and the expectation that it will eventually be deprecated in favor of a more
+well-developed spontaneous payment solution like
+[Offers](https://github.com/lightningnetwork/lightning-rfc/pull/798) or
+[AMP](https://github.com/lightningnetwork/lightning-rfc/pull/658), it makes more
+sense for keysend to be documented in a bLIP than included in the official BOLTs.
```
Comment (ariard) [inline blips/blip-0002.mediawiki]: Ultimately I think the decision to choose between the BOLT or bLIP process belong to the standard author, or at least it's a meta-discussion not belonging to the spec document itself. From years from now this section will look irrelevant to the readers.

File: blips/blip-0002.mediawiki [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+==Keysend Drawbacks==
+
+* Inability for the payee to specify their preferred `min_final_cltv_expiry`.
+  This is an issue because payer and payee may have differing security
+  requirements, which could lead to payment failures if the payee considers the
+  payer's choice of `min_final_cltv_expiry` too high or too low.
+* Multi-path payments currently not supported
+* Loss of being able to use the preimage as proof-of-payment. With non-keysend
+  payments, payers could use the payment preimage in a situation where they need
```
Comment (ariard) [inline blips/blip-0002.mediawiki]: I think the proof-of-payment property has to encompassed the knowledge of a signature for the invoice for which it is claimed the payment has been made for by the payer. Otherwise any intermediary node could show up with the relayed preimage.

**Review by ariard — COMMENTED** [review]
Good!

File: blips/blip-0002.mediawiki (hunk 7 above)
Comment (valentinewallace) [reply blips/blip-0002.mediawiki:42]: You mean TLV type `5482373484`? Maybe in [this section](https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md#tlv_payload-format)?

File: blips/blip-0002.mediawiki [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  need to prove that they made a payment (the invoice signature is necessary to
+  prove that the payer was not an intermediary hop). This proof cannot be
+  provided for keysend payments because the preimage is provided by the payer to
+  begin with and no invoice is being paid.
+* Keysend is expected to be deprecated in favor of a more well-developed
+  spontaneous payment solution like
+  [Offers](https://github.com/lightningnetwork/lightning-rfc/pull/798) or
+  [AMP](https://github.com/lightningnetwork/lightning-rfc/pull/658)
+
+==Reference implementations==
```
Comment (t-bast) [inline blips/blip-0002.mediawiki:102]: nit: you can also add eclair: https://github.com/ACINQ/eclair/pull/1485

File: 09-features.md [hunk 11]
```diff
@@ -42,6 +42,12 @@ The Context column decodes as follows:
 | 22/23 | `option_anchors_zero_fee_htlc_tx` | Anchor commitment type with zero fee HTLC transactions   | IN       |                   | [BOLT #3][bolt03-htlc-tx], [lightning-dev][ml-sighash-single-harmful]|
 | 26/27 | `option_shutdown_anysegwit`         | Future segwit versions allowed in `shutdown`              | IN       |                   | [BOLT #2][bolt02-shutdown]   |
 
+
+## bLIP Reserved Feature Bits
+| Bits  | Name                             | Description                                               | Context  | Dependencies      | Link                                  |
+|-------|----------------------------------|-----------------------------------------------------------|----------|-------------------|---------------------------------------|
+| 55    | `keysend`                        | Node supports receiving keysend payments                  | N        |                   | [bLIP #0002][blip-0002]               |
```
Comment (t-bast) [inline 09-features.md:49]: Are people relying on this specific value? It's a bit of a waste that it used a value so low, especially for a feature that will be obsolete once replaced by AMP...it would be nice if bLIPs used higher values for feature to avoid mixing official bolt features and bLIP features (but maybe that's just my OCD speaking)

**Review by t-bast — COMMENTED** [review]
Thanks, it will be useful to have this in a shared document :+1:

File: 09-features.md (hunk 11 above)
Comment (Roasbeef) [reply 09-features.md:49]: The tradeoff with higher feature bits, is that since we don't do any sparse encoding, higher feature bits (say in the tens of thousands) require the entire network to send around more data (looots of zeros onthe wire).

File: blips/blip-0002.mediawiki [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+This bLIP is dual-licensed under the Open Publication License and BSD 2-clause
+license.
+
+==Specification==
+
+Sender:
+* MUST include a TLV record keyed by type `5482373484` with a TLV value of a
+  randomly generated and cryptographically-secure 32-byte value that serves as
+  the HTLC payment preimage
+* MUST NOT set a `payment_data` field
```
Comment (Roasbeef) [inline blips/blip-0002.mediawiki:33]: What's `payment_data` here?

Comment (valentinewallace) [reply blips/blip-0002.mediawiki]: From [this section](https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md#tlv_payload-format). I clarified a bit.

**Comment by Roasbeef** [comment]
Should be moved to https://github.com/lightning/blips

**Comment by valentinewallace** [comment]
> Should be moved to https://github.com/lightning/blips

Thanks, moving!
